### PR TITLE
[downloader] Smarter updates and patch notes

### PR DESCRIPTION
**Pull request by calebj**
Cumulation of previous work done by Redjumpman and myself:
* Blocking repo updates with a swag progress indicator
  - The existing method of 'start *all* the `git pull`s and wait ten seconds' is dumb.
* Smart reload of only changed cogs
  - Reloading all repo cogs because they *might* have been updated... is dumb.
* Doesn't break when cogs are removed from a repo
  - This is also dumb.
* Display of added, removed, and updated cogs
* Patch notes for updated cogs using `git log` (original idea by @Redjumpman)

DISCLAIMER: No @irdumbs's were hurt in the making of this PR

### CODE REVIEW

File: cogs/downloader.py [hunk 1]
```diff
@@ -304,30 +383,72 @@ def get_dir_name(self, url):
...
-            # It's blocking but it shouldn't matter
-            call(["git", "clone", url, "data/downloader/" + name])
+            run(["git", "clone", url, dd + name])
         else:
-            Popen(["git", "-C", "data/downloader/" + name, "stash", "-q"])
-            Popen(["git", "-C", "data/downloader/" + name, "pull", "-q"])
+            rpcmd = ["git", "-C", dd + name, "rev-parse", "HEAD"]
+            p = run(rpcmd, stdout=PIPE)
+            oldhash = p.stdout.decode().strip()
+            run(["git", "-C", dd + name, "stash", "-q"])
```
Comment (tekulvw) [inline cogs/downloader.py]: Can we change this to `git reset --hard HEAD`?

**Comment by Jan200101** [comment]
Seems clean and good

**Review by tekulvw — CHANGES_REQUESTED** [review]

**Review by tekulvw — APPROVED** [review]

**Comment by calebj** [comment]
@Twentysix26 please test the parallel updates and see if you have the same issue as before
EDIT: oops I forgot to take out debug prints
Also don't merge until I can reconcile that pagify that got trampled over somehow, and squash

**Review by Redjumpman — APPROVED** [review]
+1

**Comment by calebj** [comment]
Needs some final testing for the windows plebs, but it's squashed and ret2merge otherwise 🎉

**Comment by calebj** [comment]
Appears to work fine on windows, I will have to add error handling for each `git` call later.

**Comment by Twentysix26** [comment]
Seems to be working fine. Want to add some finishing touches?

**Review by Jan200101 — APPROVED** [review]

**Comment by calebj** [comment]
Nah, it's good. Like I said, I'll implement detailed error handling for strange cases like already-cloned repos, filenames too long/incompatible with windows, and repos with an invalid structure in a future PR.

**Comment by Redjumpman** [comment]
No bugs that I can replicate. Downloader seems to be ready for production

**Review by Jan200101 — APPROVED** [review]
Great. Tested on Linux 4.4.12v7-aufs and Windows 10 both worked and I am already using it

**Comment by Kowlin** [comment]
Going to give this a shot on my ends, but I doubt I have any problems, will tag once its done.

**Review by YeahImRose — APPROVED** [review]
Worked with no problems on Ubuntu 16.04 for me :+1:

**Review by Kowlin — APPROVED** [review]
Core functionality seems to be working. I'll add another comment on the stuff that could use some touching up. For now I think this can be merged with some work down the line.

**Comment by Kowlin** [comment]
For the extra stuff, I doubt the bot will time out a thread if that thread requires additional information such as an authenticated `git pull`, also the bot can return a `NoneType` if the repo isn't found. But this isn't exactly new news.

**Review by tmercswims — APPROVED** [review]
Working great for me!

**Pull request merged**